File: app/src/gui/helpers/GoogleDriveSyncWorker.py
# GoogleDriveSyncWorker.py
import time
from typing import Optional
try:
    from app.src.web.google_drive_sync import GoogleDriveSync as GDS
except:
    from src.web.google_drive_sync import GoogleDriveSync as GDS

from PySide6.QtCore import QObject, Signal, QRunnable, QThreadPool, QMutex


class GoogleDriveSyncWorkerSignals(QObject):
    status_update = Signal(str)
    sync_finished = Signal(bool, str)


class GoogleDriveSyncWorker(QRunnable):
    def __init__(self, 
                 key_file: str, 
                 local_path: str, 
                 remote_path: str, 
                 dry_run: bool, 
                 user_email_to_share_with: Optional[str] = None # NEW PARAMETER
    ):
        super().__init__()
        self.key_file = key_file
        self.local_path = local_path
        self.remote_path = remote_path
        self.dry_run = dry_run
        self.share_email = user_email_to_share_with # NEW ATTRIBUTE
        self.signals = GoogleDriveSyncWorkerSignals()
        self._is_running = True

    def _log(self, message: str):
        if self._is_running:
            timestamp = time.strftime("[%H:%M:%S]")
            self.signals.status_update.emit(f"{timestamp} {message}")

    def run(self):
        self.signals.status_update.emit("\n" + "="*50)
        self._log(f"--- Google Drive Sync Initiated ---")
        self._log(f"Mode: {'DRY RUN' if self.dry_run else 'LIVE'}")
        self.signals.status_update.emit("="*50 + "\n")

        try:
            sync_manager = GDS(
                service_account_file=self.key_file,
                local_source_path=self.local_path,
                drive_destination_folder_name=self.remote_path,
                dry_run=self.dry_run,
                logger=self._log,
                # Pass new parameter to GDS
                user_email_to_share_with=self.share_email
            )
            success, final_message = sync_manager.execute_sync()
        except Exception as e:
            success = False
            final_message = f"Critical error: {e}"
            self._log(f"ERROR: {final_message}")

        # Only emit if still alive
        if self._is_running:
            self.signals.sync_finished.emit(success, final_message)

    def stop(self):
        self._is_running = False

File: app/src/gui/helpers/GoogleDriveSyncWorker.cpp
#include "GoogleDriveSyncWorker.h"
#include "web/google_drive_sync.h" // Assumed C++ equivalent
#include <QDateTime>
#include <QPair> // For the return value

GoogleDriveSyncWorker::GoogleDriveSyncWorker(
    const QVariantMap& authConfig,
    const QString& localPath, 
    const QString& remotePath, 
    bool dryRun, 
    const QString& userEmailToShareWith,
    QObject* parent
)
    : QObject(parent),
      m_authConfig(authConfig),
      m_localPath(localPath),
      m_remotePath(remotePath),
      m_dryRun(dryRun),
      m_shareEmail(userEmailToShareWith),
      m_isRunning(true)
{
    m_authMode = m_authConfig.value("mode", "unknown").toString();
}

void GoogleDriveSyncWorker::log(const QString& message)
{
    if (m_isRunning) {
        QString timestamp = QDateTime::currentDateTime().toString("[HH:mm:ss]");
        emit statusUpdate(QString("%1 %2").arg(timestamp, message));
    }
}

void GoogleDriveSyncWorker::run()
{
    emit statusUpdate("\n" + QString("=").repeated(50));
    log(QString("--- Google Drive Sync Initiated ---"));
    log(QString("Authentication Mode: %1").arg(m_authMode.toUpper()));
    log(QString("Sync Mode: %1").arg(m_dryRun ? "DRY RUN" : "LIVE"));
    emit statusUpdate(QString("=").repeated(50) + "\n");

    bool success = false;
    QString finalMessage = "Cancelled by user.";
    GoogleDriveSync* syncManager = nullptr;
    
    // Logger function (C++ lambda)
    auto logger = [this](const QString& msg){ this->log(msg); };

    try {
        if (m_authMode == "service_account") {
            syncManager = new GoogleDriveSync(
                m_localPath,
                m_remotePath,
                m_dryRun,
                logger,
                m_authConfig.value("key_file").toString(),
                m_shareEmail
            );
        } else if (m_authMode == "personal_account") {
             syncManager = new GoogleDriveSync(
                m_localPath,
                m_remotePath,
                m_dryRun,
                logger,
                m_authConfig.value("client_secrets_file").toString(),
                m_authConfig.value("token_file").toString()
            );
        } else {
             throw std::runtime_error(QString("Unsupported authentication mode: %1").arg(m_authMode).toStdString());
        }

        if (m_isRunning) {
            QPair<bool, QString> result = syncManager->execute_sync();
            success = result.first;
            finalMessage = result.second;
        }
        
    } catch (const std::exception& e) {
        success = false;
        finalMessage = QString("Critical error: %1").arg(e.what());
        log(QString("ERROR: %1").arg(finalMessage));
    }
    
    delete syncManager; // Clean up
    syncManager = nullptr;

    if (!m_isRunning && success) {
        success = false;
        finalMessage = "Synchronization manually cancelled.";
    }

    if (m_isRunning) {
        emit syncFinished(success, finalMessage);
    }
}

void GoogleDriveSyncWorker::stop()
{
    if (m_isRunning) {
        m_isRunning = false;
        emit statusUpdate("\n!!! SYNCHRONIZATION INTERRUPTED !!!");
    }
}